### ISSUE TITLE
THNN

### ISSUE THREAD

**Issue by andreaskoepf** [torch/nn#506]
@fmassa @apaszke and I started working on a FFI based functional lua-free C-backend for torch/nn. The development takes place in [this fork](https://github.com/fmassa/nn/tree/THNN) in the THNN branch, the THNN lib resides in lib/THNN. Up to now 28 of the 44 C modules in nn have been converted (CPU only). If somebody wants to help, please feel free to contact us via [gitter](https://gitter.im/apaszke/THNN), some [style-guidelines and description of the conversion steps](https://github.com/fmassa/nn/blob/THNN/lib/THNN/README.md) exist.

We came to the conclusion that creating THNN as one big PR for nn/cunn might not be the best solution since it takes some time (maybe until X-mas) to convert all CPU/GPU modules and we want to avoid further project drift to the main nn/cunn repo since all changes to nn/cunn had to be tracked and integrated into THNN as well. Technically it is be possible to do the conversion to FFI at the granularity of one module.

We would like to discuss the following points with the torch owners (@soumith et al.):

1. Do you think it would be acceptable to have the C-backend for torch split into two libraries during a short period of time (1-2 months)?
2. From a testing perspective it would be nice to run cross-checks against the old nn code. But for contributors who want to send in fixes having the old impls flying around could be confusing. Also to keep the git-history of the C impls it would be best to use git-mv from nn/generic to nn/lib/THNN/generic. If we remove the original impl we would have to find a way to run the old C-backend vs. THNN functional API tests externally.
3. We have planned to use a vtable-lookup by a tensor:type() string, creating the target signature dynamically once and caching the function returned by ffi. The 'old' nn impl uses the metatable of the tensors to store the nn-function-table. Should we keep the old metatable-misuse or go for the tensor-type-vtable lookup-by string?
4. Some modules need access to a random generator. Generating random numbers differs for cpu/cuda (e.g. THRandom_uniform vs. curand_uniform). Is it important to be able to pass in a custom random generator (e.g. control rgn-state/seed on a call-by-call basis etc.) or could backend-specific global rngs be part of the THNNState* which is now passed as first arg to each backend-function?

Also there are some more general questions:
A reason for creating THNN is to be able to reference the torch functions from a C/C++ library that would mirror the lua side, e.g. to allow use of torch models in production code without lua dependency (e.g. no need to break out of the torch universe and  use Caffe). 
Which models are the most important? 
Who would be interested in using a pure-c version (maybe tell us a bit about your expectations, use-case, requirements)?

**Comment by hughperkins** [comment]
Interesting.  You mean, all `nn` modules will be able to be used without needing lua?  From the point of view of making torch nn portable, that seems to be quite good news, ie can wrap in python/julia/lua/java/whatever/one/wants.

> We have planned to use a vtable-lookup by a tensor:type() string, creating the target signature dynamically once and caching the function returned by ffi.

I havent fully understood this.  Just to check, this is only needed when calling from lua into the c library right?  No dynamic stuff needed if calling from c/c++ right?

Personal preference for using C++, rather than C, but if the price to pay for using C++ is that it becomes a header-only template library, I'd rather choose C :-)

**Comment by apaszke** [comment]
@hughperkins Don't worry, nothing dynamic needed in C/C++ :blush: The vtable is only needed for wrapping THNN in Lua. C functions are generated in `generic` folder, so you have to choose one that corresponds to your tensor type. Current nn solution that puts them into Tensor metatables isn't really intuitive. 

Also, the plan is that THNN will have pure functional API. There will be no modules/layers/objects in it, just a set of functions like that one:
```C
THNN_FloatLinear_updateOutput(
          THNNState *state,
          THFloatTensor *input,
          THFloatTensor *output,
          THFloatTensor *weight
          THFloatTensor *bias)
```
So we're basically just making a lua-independent C implementation that at the moment requires mostly refactoring out all the lua stuff from `generic` in nn.

When THNN will be finished we'll start writing another C/C++ library, that will be object oriented and will resemble nn. It will wrap THNN in a way that will be more convenient than calling functions with 6 arguments all the time and enable quick embedding of trained networks into other projects.

I really love the idea of a python wrapper! I'm thinking about it from time to time, but I gues I'll wait for the THNN wrapper to be finished. It should be much simpler then, because you won't have to worry about all the lua stuff.

**Comment by kirill-pinigin** [comment]
When the C++ Wrapper for torch/nn will be ready?

**Comment by fmassa** [comment]
@kirill-pinigin There is a C++ wrapper for TH in https://github.com/pytorch/pytorch/tree/master/torch/lib/THPP , but not for THNN.

**Comment by kirill-pinigin** [comment]
and what is about THNN ?   Do you know?

**Comment by fmassa** [comment]
@kirill-pinigin maybe check https://github.com/zdevito/ATen
